feugy/tabulous
# feat(web): first implementation of game replay

### :book: What's in there?

Multiplayer board games, especially when played asynchronously, need a replay function, so one can see what other players did.
This game history is mixed together with the players discussion, to give a bit of context to the exchanged messages.

While replaying, a player can not interact any more with meshes (no selection, moves, actions, menu, details), but has still full control of the camera.
Peer players can replay independently from each other, and continue playing while other peers are replaying.

Are included here:

- fix(web): wrong angle when rotating multiple stacked card.
- fix(web): wrong elevation when dragging a mah-jong tile from hand.
- fix(klondike): some cards are missing symbols.
- feat(server): supports player actions and moves history.
- feat(web): brings a first implementation for game replay.
- feat(klondike): automatically flips first discarded card.
- refactor(server): introduces play action to distinguish draws.
- chore: updates all dependencies
- chore: patches vitest to fix OOM errors when diffing meshes.

### :test_tube: How to test?

_Wrong angle_
1. create a new "playground" game with cards
   > It shows the game with a stack of cards
1. right click on the card 
   > It opens the action menu with a rotate quantity button
1. rotate one (or many) cards
   > Their rotation is 45° (it used to be broken:)

   ![image](https://github.com/feugy/tabulous/assets/186268/16efa714-cbc5-4cdb-b2c8-5f26bda50663)

_Wrong elevation_
1. create a new mah-jong" game with any player color
   > It shows the game with "walls" of tiles
1. Right click on a tile and draw it in your hand
   > the tile disappear from the table to appear in your hand
1. drag your tile to the table
   > as it leaves your hand to join the table, the tile elevation is above the table (it used to be _partially below_ it)

   [Capture vidéo du 2023-09-16 17-01-09.webm](https://github.com/feugy/tabulous/assets/186268/b2c30370-926d-4b1c-bfa7-eb629992eada)

_Missing symbols_
1. Create a new klondike game
   > All 10, 9, 8, 7 cards have their symbols (some uses to be missing or misplaced)

   ![image](https://github.com/feugy/tabulous/assets/186268/2d18b4e3-4e0e-4716-bff1-88d8f2a6b06e)

_Replay_

Enjoy this tiny video with a mix of actions, and actions in cascade (draw and unsnap)
[Capture vidéo du 2023-09-16 17-26-51.webm](https://github.com/feugy/tabulous/assets/186268/1b10e0dc-ed04-4365-b458-c84db23cff34)



### :up: Notes to reviewers

This implementation has a big flaw: when the host is replaying, its main game engine does not update any more when receiving peer actions, which means the game is not saved any more, and new peer would receive incomplete game.
